File: src/ovi_voice_assistant/scheduler.py
"""Scheduler — persistent cron-based automations that run agent prompts proactively."""

import asyncio
import datetime
import json
import logging
import uuid
from collections.abc import Awaitable, Callable
from dataclasses import asdict, dataclass
from pathlib import Path

logger = logging.getLogger(__name__)

# How often the scheduler checks for automations to fire.
_CHECK_INTERVAL_S = 30


@dataclass
class Automation:
    """A single scheduled automation."""

    id: str
    name: str
    schedule: str  # cron expression: minute hour dom month dow
    prompt: str  # what to ask the agent
    enabled: bool = True
    last_run: str = ""  # ISO timestamp — prevents double-firing within same minute


# ---------------------------------------------------------------------------
# Cron matching (no external dependency)
# ---------------------------------------------------------------------------


def _matches_field(field_expr: str, value: int) -> bool:
    """Check if a single cron field matches a value."""
    if field_expr == "*":
        return True
    for part in field_expr.split(","):
        if "/" in part:
            base, step_s = part.split("/", 1)
            step = int(step_s)
            start = 0 if base == "*" else int(base)
            if (value - start) >= 0 and (value - start) % step == 0:
                return True
        elif "-" in part:
            lo, hi = part.split("-", 1)
            if int(lo) <= value <= int(hi):
                return True
        else:
            if int(part) == value:
                return True
    return False


def cron_matches(expr: str, dt: datetime.datetime) -> bool:
    """Check if a 5-field cron expression matches a datetime.

    Fields: minute hour day-of-month month day-of-week
    Day-of-week: 0 = Sunday, 6 = Saturday (standard cron convention).
    """
    parts = expr.strip().split()
    if len(parts) != 5:
        return False
    minute, hour, dom, month, dow = parts
    # Python isoweekday: 1=Mon..7=Sun → cron: 0=Sun, 1=Mon..6=Sat
    cron_dow = dt.isoweekday() % 7
    return (
        _matches_field(minute, dt.minute)
        and _matches_field(hour, dt.hour)
        and _matches_field(dom, dt.day)
        and _matches_field(month, dt.month)
        and _matches_field(dow, cron_dow)
    )


# ---------------------------------------------------------------------------
# Scheduler
# ---------------------------------------------------------------------------

# Callback types for the scheduler.
RunPrompt = Callable[[str], Awaitable[str]]
AnnounceAll = Callable[[str], Awaitable[None]]


class Scheduler:
    """Runs persistent cron automations in the background.

    Each automation has a cron schedule and a prompt.  When the schedule
    fires, the prompt is sent to the agent and the response is announced
    on all connected devices.
    """

    def __init__(
        self,
        path: Path,
        run_prompt: RunPrompt,
        announce: AnnounceAll,
    ) -> None:
        self._path = path
        self._run_prompt = run_prompt
        self._announce = announce
        self._automations: list[Automation] = []
        self._task: asyncio.Task | None = None
        self._fire_tasks: set[asyncio.Task] = set()

    # -- Persistence --

    def load(self) -> None:
        """Load automations from disk."""
        if self._path.exists():
            try:
                data = json.loads(self._path.read_text())
                self._automations = [Automation(**a) for a in data]
                logger.info(
                    "Loaded %d automation(s) from %s",
                    len(self._automations),
                    self._path,
                )
            except Exception:
                logger.exception("Failed to load automations from %s", self._path)
                self._automations = []
        else:
            self._automations = []

    def _save(self) -> None:
        """Persist automations to disk."""
        self._path.parent.mkdir(parents=True, exist_ok=True)
        self._path.write_text(
            json.dumps([asdict(a) for a in self._automations], indent=2)
        )

    # -- Public API (called by agent tools) --

    @property
    def automations(self) -> list[Automation]:
        return list(self._automations)

    def create(self, name: str, schedule: str, prompt: str) -> Automation:
        """Create and persist a new automation."""
        # Validate cron expression
        parts = schedule.strip().split()
        if len(parts) != 5:
            raise ValueError(
                f"Invalid cron expression: {schedule!r}. "
                "Expected 5 fields: minute hour day-of-month month day-of-week"
            )
        auto = Automation(
            id=uuid.uuid4().hex[:8],
            name=name,
            schedule=schedule,
            prompt=prompt,
        )
        self._automations.append(auto)
        self._save()
        logger.info("Automation created: %s (%s) → %r", name, schedule, prompt[:60])
        return auto

    def delete(self, name: str) -> bool:
        """Delete an automation by name. Returns True if found."""
        before = len(self._automations)
        self._automations = [a for a in self._automations if a.name != name]
        if len(self._automations) < before:
            self._save()
            logger.info("Automation deleted: %s", name)
            return True
        return False

    def set_enabled(self, name: str, enabled: bool) -> bool:
        """Enable or disable an automation by name. Returns True if found."""
        for a in self._automations:
            if a.name == name:
                a.enabled = enabled
                self._save()
                logger.info(
                    "Automation %s: %s", "enabled" if enabled else "disabled", name
                )
                return True
        return False

    # -- Background loop --

    async def start(self) -> None:
        self._task = asyncio.create_task(self._loop(), name="scheduler")
        logger.info("Scheduler started (%d automations)", len(self._automations))

    async def stop(self) -> None:
        if self._task:
            self._task.cancel()
            try:
                await self._task
            except (asyncio.CancelledError, Exception):
                pass
        logger.info("Scheduler stopped")

    async def _loop(self) -> None:
        while True:
            try:
                await self._check()
            except asyncio.CancelledError:
                raise
            except Exception:
                logger.exception("Scheduler check error")
            await asyncio.sleep(_CHECK_INTERVAL_S)

    async def _check(self) -> None:
        now = datetime.datetime.now().astimezone()
        now_minute = now.replace(second=0, microsecond=0)

        for auto in self._automations:
            if not auto.enabled:
                continue
            if not cron_matches(auto.schedule, now):
                continue
            # Prevent double-firing within the same minute.
            if auto.last_run:
                last = datetime.datetime.fromisoformat(auto.last_run)
                if last.replace(second=0, microsecond=0) >= now_minute:
                    continue
            auto.last_run = now.isoformat()
            self._save()
            task = asyncio.create_task(
                self._fire(auto),
                name=f"automation-{auto.name}",
            )
            self._fire_tasks.add(task)
            task.add_done_callback(self._fire_tasks.discard)

    async def _fire(self, auto: Automation) -> None:
        """Run the automation's prompt through the agent and announce the result."""
        logger.info("Automation firing: %s → %r", auto.name, auto.prompt[:60])
        try:
            response = await self._run_prompt(auto.prompt)
            if response.strip():
                await self._announce(response)
                logger.info("Automation announced: %s → %r", auto.name, response[:80])
            else:
                logger.warning("Automation %s produced empty response", auto.name)
        except Exception:
            logger.exception("Automation %s failed", auto.name)

File: internal/scheduler/scheduler.go
// Package scheduler runs persistent cron-based automations that send
// agent prompts proactively and announce the result.
package scheduler

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"log/slog"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// CheckInterval is how often the scheduler checks for automations to fire.
const CheckInterval = 30 * time.Second

// Automation is a single scheduled automation.
type Automation struct {
	ID       string `json:"id"`
	Name     string `json:"name"`
	Schedule string `json:"schedule"` // cron expression: minute hour dom month dow
	Prompt   string `json:"prompt"`   // what to ask the agent
	Enabled  bool   `json:"enabled"`
	LastRun  string `json:"last_run"` // RFC3339 — prevents double-firing within same minute
}

// RunPrompt runs a prompt through the agent and returns the response.
type RunPrompt func(ctx context.Context, prompt string) (string, error)

// Announce speaks text on all devices.
type Announce func(ctx context.Context, text string) error

// Scheduler runs persistent cron automations in the background.
type Scheduler struct {
	path      string
	runPrompt RunPrompt
	announce  Announce

	mu          sync.Mutex
	automations []Automation
	cancel      context.CancelFunc
	loopDone    chan struct{}
	fires       sync.WaitGroup

	// Now is the clock; tests may override it.
	Now func() time.Time
}

// New creates a scheduler persisting to path.
func New(path string, runPrompt RunPrompt, announce Announce) *Scheduler {
	return &Scheduler{path: path, runPrompt: runPrompt, announce: announce, Now: time.Now}
}

// -- Persistence --

// Load reads automations from disk. A missing or corrupt file yields an
// empty list.
func (s *Scheduler) Load() {
	s.mu.Lock()
	defer s.mu.Unlock()
	data, err := os.ReadFile(s.path)
	if err != nil {
		s.automations = nil
		return
	}
	var autos []Automation
	if err := json.Unmarshal(data, &autos); err != nil {
		slog.Error("Failed to load automations", "path", s.path, "err", err)
		s.automations = nil
		return
	}
	s.automations = autos
	slog.Info("Loaded automations", "count", len(autos), "path", s.path)
}

// save persists automations. Caller must hold mu.
func (s *Scheduler) save() {
	if err := os.MkdirAll(filepath.Dir(s.path), 0o755); err != nil {
		slog.Error("Failed to create automations dir", "err", err)
		return
	}
	autos := s.automations
	if autos == nil {
		autos = []Automation{}
	}
	data, err := json.MarshalIndent(autos, "", "  ")
	if err != nil {
		return
	}
	if err := os.WriteFile(s.path, data, 0o644); err != nil {
		slog.Error("Failed to save automations", "err", err)
	}
}

// -- Public API (called by agent tools) --

// Automations returns a copy of the current list.
func (s *Scheduler) Automations() []Automation {
	s.mu.Lock()
	defer s.mu.Unlock()
	return append([]Automation(nil), s.automations...)
}

// Create validates, persists and returns a new automation.
func (s *Scheduler) Create(name, schedule, prompt string) (Automation, error) {
	if !ValidateCron(schedule) {
		return Automation{}, fmt.Errorf(
			"invalid cron expression %q: expected 5 fields (minute hour day-of-month month day-of-week)",
			schedule)
	}
	auto := Automation{
		ID:       newID(),
		Name:     name,
		Schedule: schedule,
		Prompt:   prompt,
		Enabled:  true,
	}
	s.mu.Lock()
	s.automations = append(s.automations, auto)
	s.save()
	s.mu.Unlock()
	slog.Info("Automation created", "name", name, "schedule", schedule, "prompt", truncate(prompt, 60))
	return auto, nil
}

// Delete removes an automation by name. Returns true if found.
func (s *Scheduler) Delete(name string) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	kept := s.automations[:0:0]
	for _, a := range s.automations {
		if a.Name != name {
			kept = append(kept, a)
		}
	}
	if len(kept) == len(s.automations) {
		return false
	}
	s.automations = kept
	s.save()
	slog.Info("Automation deleted", "name", name)
	return true
}

// SetEnabled enables or disables an automation by name. Returns true if found.
func (s *Scheduler) SetEnabled(name string, enabled bool) bool {
	s.mu.Lock()
	defer s.mu.Unlock()
	for i := range s.automations {
		if s.automations[i].Name == name {
			s.automations[i].Enabled = enabled
			s.save()
			slog.Info("Automation toggled", "name", name, "enabled", enabled)
			return true
		}
	}
	return false
}

// -- Background loop --

// Start launches the periodic check loop.
func (s *Scheduler) Start() {
	s.mu.Lock()
	defer s.mu.Unlock()
	if s.cancel != nil {
		return
	}
	ctx, cancel := context.WithCancel(context.Background())
	s.cancel = cancel
	s.loopDone = make(chan struct{})
	go s.loop(ctx, s.loopDone)
	slog.Info("Scheduler started", "automations", len(s.automations))
}

// Stop halts the loop and waits for in-flight automations to finish.
func (s *Scheduler) Stop() {
	s.mu.Lock()
	cancel := s.cancel
	done := s.loopDone
	s.cancel = nil
	s.loopDone = nil
	s.mu.Unlock()
	if cancel != nil {
		cancel()
		<-done
	}
	s.fires.Wait()
	slog.Info("Scheduler stopped")
}

func (s *Scheduler) loop(ctx context.Context, done chan struct{}) {
	defer close(done)
	ticker := time.NewTicker(CheckInterval)
	defer ticker.Stop()
	for {
		s.Check(ctx)
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
		}
	}
}

// Check fires every enabled automation whose schedule matches the current
// minute and has not already fired in that minute.
func (s *Scheduler) Check(ctx context.Context) {
	now := s.Now()
	nowMinute := now.Truncate(time.Minute)

	s.mu.Lock()
	var toFire []Automation
	for i := range s.automations {
		a := &s.automations[i]
		if !a.Enabled || !CronMatches(a.Schedule, now) {
			continue
		}
		if a.LastRun != "" {
			if last, err := time.Parse(time.RFC3339Nano, a.LastRun); err == nil &&
				!last.Truncate(time.Minute).Before(nowMinute) {
				continue
			}
		}
		a.LastRun = now.Format(time.RFC3339Nano)
		toFire = append(toFire, *a)
	}
	if len(toFire) > 0 {
		s.save()
	}
	s.mu.Unlock()

	for _, a := range toFire {
		s.fires.Add(1)
		go func(a Automation) {
			defer s.fires.Done()
			s.Fire(ctx, a)
		}(a)
	}
}

// Fire runs the automation's prompt through the agent and announces the result.
func (s *Scheduler) Fire(ctx context.Context, a Automation) {
	slog.Info("Automation firing", "name", a.Name, "prompt", truncate(a.Prompt, 60))
	response, err := s.runPrompt(ctx, a.Prompt)
	if err != nil {
		slog.Error("Automation failed", "name", a.Name, "err", err)
		return
	}
	if strings.TrimSpace(response) == "" {
		slog.Warn("Automation produced empty response", "name", a.Name)
		return
	}
	if err := s.announce(ctx, response); err != nil {
		slog.Error("Automation announce failed", "name", a.Name, "err", err)
		return
	}
	slog.Info("Automation announced", "name", a.Name, "response", truncate(response, 80))
}

func newID() string {
	b := make([]byte, 4)
	_, _ = rand.Read(b)
	return hex.EncodeToString(b)
}

func truncate(s string, n int) string {
	r := []rune(s)
	if len(r) <= n {
		return s
	}
	return string(r[:n])
}
